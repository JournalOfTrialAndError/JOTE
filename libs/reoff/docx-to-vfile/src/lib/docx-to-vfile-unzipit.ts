import { VFile } from 'vfile'
import { unzip } from 'unzipit'

const removeHeader = (text: string | undefined) =>
  text ? text.replace(/<\?xml.*?\?>/, '') : ''

export async function docxToVFile(file: ArrayBuffer) {
  const { entries } = await unzip(file)
  console.log(entries['word/document.xml'])
  const doc = (await entries['word/document.xml'].arrayBuffer()).toString()
  const foot =
    (await entries?.['word/footnotes.xml']?.arrayBuffer())?.toString() || ''
  const bib =
    (await entries?.['customXml/item1.xml']?.arrayBuffer())?.toString() || ''
  // const data = await getXMLDatas(file, {
  //   filenames: [
  //     /customXml\/item\d+\.xml/,
  //     'word/document.xml',
  //     'word/footnotes.xml',
  //   ],
  // })

  // const {
  //   'word/document.xml': document,
  //   'word/footnotes.xml': footnotes,
  //   ...bibliography
  // } = data
  const total = `${removeHeader(doc).slice(0, -'</w:document>'.length)}
  ${removeHeader(foot)}
  ${Object.values(bib)
    .map((bib) => removeHeader(bib))
    .join('\n')}
  </w:document>`
  const vfile = new VFile(total)
  // if (footnotes) {
  //   Object.assign(vfile.data, { footnotes })
  // }
  return vfile
}
