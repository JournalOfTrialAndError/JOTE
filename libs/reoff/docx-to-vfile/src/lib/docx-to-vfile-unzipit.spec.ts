import { docxToVFile } from './docx-to-vfile-unzipit'
import fs from 'fs'
import path from 'path'

describe('reoffDocxToVfile', () => {
  const doc = fs.readFileSync(
    path.join(__dirname, '../../../reoff-parse/src/test/word-citation.docx')
  )
  jest.setTimeout(10000)
  it('should work', async () => {
    const vfile = await docxToVFile(new Uint8Array(doc))
    console.log(vfile)
    expect(vfile).toMatchSnapshot()
  })
})
