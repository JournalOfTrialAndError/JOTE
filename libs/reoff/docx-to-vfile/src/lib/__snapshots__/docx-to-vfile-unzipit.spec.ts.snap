// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reoffDocxToVfile should work 1`] = `
VFile {
  "cwd": "/Users/thomas/Projects/jote-monorepo",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
  "value": "[object
  
  [
o
b
j
e
c
t
 
A
r
r
a
y
B
u
f
f
e
r
]
  </w:document>",
}
`;
